### PR TITLE
Variables: turns on newVariables as a new default

### LINKED ISSUE #22564: Template Variables move to React/Redux Leftovers
**Issues to be fixed:**
- [x] Dashboard refresh does no longer work. The path there (in current variable system) -> refresh -> event timeRangeUpdated -> variableSrv updates any variables -> dashboard.templateVariablesUpdated -> dashboad.startRefresh() #22821
- [x] Single value variables should close and commit Picker when value is selected. #22823
- [x] Preserve order of variables when saving dashboard #22805
- [x] Infinite loop when using StackDriver #22917 (fixed by @daniellee)
- [x] Issue with ["value"] in current for single value variable #23208
- [x] Ad hoc filter not applied after page reload (from url) (#23106) (wrongly configured variable) 
- [x] Issue with updating query variable regex (template_srv - reading from getState). #23229

**Bare minimum**
- [x] Introduce data source variable type #22770
- [x] Introduce ad hoc variable type #22784
- [x] Introduce interval variable type #22797
- [x] Create proper dashboard migration step #23107 
- [ ] Remove feature toggle

**PR comments**
- [x] Remove uuid _Originally posted by @torkelo in (https://github.com/grafana/grafana/pull/22434_) #22858
- [x] Use registry instead of VariableAdapters. _Originally posted by @hugohaggmark in (https://github.com/grafana/grafana/pull/22434/review_comment/create_) #22866
- [x] Maybe this file could be split to variable_utils and types, as it's quite big? _Originally posted by @dprokop in (https://github.com/grafana/grafana/pull/22434_) #22884

**Misc**
- [ ] #20703
- [ ] New Form styles
- [ ] #22254
- [ ] documentation
- [ ] celebrate 🎉 

**When we have nothing else to do:**
- [ ] Create a tag component to handle the picker tags
- [ ] Change so the VariableTag and VariableOption uses text/value: string[] instead of string | string[]
- [x] Maybe move all the new React/Redux files to public/app/features/variables
- [ ] Have a look a see if we can make the adapter interface a bit simpler. Some of the functionality is shared between all variables.
- [ ] Move pickers folder into `public/app/features/variables/shared`. 
- [ ] Move editors folder into `public/app/features/variables/shared`. 
- [ ] Split OptionsPicker into one with multi support and one without.

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
Turns on newVariables feature as a new default. My thinking is we run with this in master for a while before cleaning up the old system.

**Which issue(s) this PR fixes**:
Closes #22564

**Special notes for your reviewer**:

### CODE REVIEW

**Review by mckn — APPROVED** [review]
LGTM

File: public/app/features/templating/template_srv.ts [hunk 1]
```diff
@@ -419,15 +431,15 @@ export class TemplateSrv {
...
+      return this.dependencies.getVariableWithName(name);
     }
 
     return this.index[name];
   };
 
   private getAdHocVariables = (): any[] => {
     if (getConfig().featureToggles.newVariables) {
-      return getFilteredVariables(isAdHoc);
+      return this.dependencies.getFilteredVariables(isAdHoc);
```
Comment (hugohaggmark) [inline public/app/features/templating/template_srv.ts:442]: Just a thought, this will be in unsorted order.

Comment (mckn) [reply public/app/features/templating/template_srv.ts:442]: hmm - good point. Need to have a look at where it is used.

**Comment by torkelo** [comment]
One question that just maybe has been missed: variableSrv on init set's two global variables __dashboard & __org  
https://github.com/grafana/grafana/blob/master/public/app/features/templating/variable_srv.ts#L56

think we need to add that somewhere to the new init process

**Comment by mckn** [comment]
> One question that just maybe has been missed: variableSrv on init set's two global variables __dashboard & __org
> https://github.com/grafana/grafana/blob/master/public/app/features/templating/variable_srv.ts#L56
> 
> think we need to add that somewhere to the new init process

Super, will have a look at that one as well.

**Review by torkelo — APPROVED** [review]
Looks good, agree with the dependency injection solution

File: public/app/features/variables/state/actions.ts [hunk 2]
```diff
@@ -72,12 +74,38 @@ export const initDashboardTemplating = (list: VariableModel[]): ThunkResult<void
...
+
     for (let index = 0; index < getVariables(getState()).length; index++) {
       dispatch(addInitLock(toVariablePayload(getVariables(getState())[index])));
     }
   };
 };
 
+export const completeDashboardTemplating = (dashboard: DashboardModel): ThunkResult<void> => {
+  return (dispatch, getState) => {
+    templateSrv.updateIndex();
```
Comment (hugohaggmark) [inline public/app/features/variables/state/actions.ts:88]: Hmm, will this have any effect as we haven't called templateSrv.init(variables)? We should have a short sync about this I think?

File: public/app/features/variables/state/actions.ts [hunk 3]
```diff
@@ -106,15 +134,18 @@ export const processVariableDependencies = async (variable: VariableModel, state
...
+): ThunkResult<Promise<void>> => {
   return async (dispatch, getState) => {
     const variable = getVariable(identifier.id!, getState());
     await processVariableDependencies(variable, getState());
 
     const urlValue = queryParams['var-' + variable.name];
     if (urlValue !== void 0) {
       await variableAdapters.get(variable.type).setValueFromUrl(variable, urlValue ?? '');
-      await dispatch(resolveInitLock(toVariablePayload(variable)));
+      dispatch(resolveInitLock(toVariablePayload(variable)));
```
Comment (hugohaggmark) [inline public/app/features/variables/state/actions.ts:149]: Awesome, nice catch 🏅

File: public/app/features/variables/state/actions.ts (hunk 2 above)
Comment (mckn) [reply public/app/features/variables/state/actions.ts:88]: hmm, this one can be skipped, I think. Good point! We don't need the index for the variabels. Only  for the globals and time.

File: public/app/features/templating/template_srv.ts [hunk 4]
```diff
@@ -25,7 +37,7 @@ export class TemplateSrv {
   private timeRange?: TimeRange | null = null;
   private fieldAccessorCache: FieldAccessorCache = {};
 
-  constructor() {
+  constructor(private dependencies: TemplateSrvDependencies = runtimeDependencies) {
```
Comment (dprokop) [inline public/app/features/templating/template_srv.ts:40]: 👍

**Review by dprokop — APPROVED** [review]
Looks good!

**Pull request merged**